crates/noirc_frontend/src/hir/def_collector/dc_crate.rs

```diff
@@ -228,6 +229,18 @@ impl DefCollector {
             def_collector.collected_traits_impls,
             errors,
         );
+        // check that a function is calling a pub function if from another module.
+        let mut visibility = FunctionVisibility::default();
+        for (_, func_id) in &file_func_ids {
+            visibility.check_visibility(context, func_id);
+        }
+        for (_, func_id) in &file_method_ids {
+            visibility.check_visibility(context, func_id);
+        }
+        for (_, func_id) in &file_trait_impls_ids {
+            visibility.check_visibility(context, func_id);
+        }
+        errors.extend(visibility.errors);
```

Comment:
We don't need a separate pass to go through each function, collect all function ids that are called, and then check the visibility on those. We can do it within the name resolution pass when these variables are referred to here: https://github.com/noir-lang/noir/blob/master/crates/noirc_frontend/src/hir/resolution/resolver.rs#L1035

Reply:
I can't retrieve the function metadata from resolver.rs#L1035: `all function ids should have metadata`, which makes sense to me because we are resolving functions and we do not guarantee it will be done in the correct order.

Reply:
Can we add metadata or just whether a function is `pub` or not during def collection instead then? It'd make sense if we populated each FuncMeta during def collection then just had to check them during resolution.

Reply:
I'm working on a branch now to move the pub check to name resolution instead. Since separating out which fields of FuncMeta we can resolve early from those we cant would be a larger change, I've just added a separate hash map now for mapping functions to their visibilities which is filled out early.